File: src/vantage/services/watcher.py
import asyncio
import logging
import os
import threading
import time
from dataclasses import dataclass
from pathlib import Path
from typing import override

from watchfiles import Change, DefaultFilter, watch

from vantage.services.socket_manager import manager
from vantage.settings import get_daemon_config, settings

logger = logging.getLogger(__name__)


def _read_inotify_limit(name: str) -> int | None:
    """Read an inotify sysctl from /proc.  Returns None on non-Linux or errors."""
    try:
        with open(f"/proc/sys/fs/inotify/{name}") as f:
            return int(f.read().strip())
    except (OSError, ValueError):
        return None


def _count_inotify_usage() -> tuple[int, int] | None:
    """Return (instances, watches) held by the current process.

    Walks ``/proc/self/fd`` looking for ``anon_inode:inotify`` links, then
    counts ``inotify wd:`` lines in each corresponding fdinfo file.  Returns
    None on non-Linux or if /proc is unavailable.
    """
    try:
        fd_dir = "/proc/self/fd"
        entries = os.listdir(fd_dir)
    except OSError:
        return None

    instances = 0
    watches = 0
    for fd in entries:
        try:
            link = os.readlink(os.path.join(fd_dir, fd))
        except OSError:
            continue
        if link != "anon_inode:inotify":
            continue
        instances += 1
        try:
            with open(f"/proc/self/fdinfo/{fd}") as f:
                for line in f:
                    if line.startswith("inotify wd:"):
                        watches += 1
        except OSError:
            continue
    return instances, watches


def _log_inotify_limits() -> None:
    """Log inotify sysctl limits (once, at watcher startup)."""
    max_watches = _read_inotify_limit("max_user_watches")
    max_instances = _read_inotify_limit("max_user_instances")
    if max_watches is None and max_instances is None:
        logger.info("[watcher] inotify limits unavailable (non-Linux or /proc not mounted)")
        return
    logger.info(
        "[watcher] inotify limits: max_user_watches=%s max_user_instances=%s",
        max_watches if max_watches is not None else "?",
        max_instances if max_instances is not None else "?",
    )


def _log_inotify_usage(context: str) -> None:
    """Log current-process inotify usage and warn if close to the ceiling."""
    usage = _count_inotify_usage()
    if usage is None:
        return
    instances, watches = usage
    max_watches = _read_inotify_limit("max_user_watches")
    max_instances = _read_inotify_limit("max_user_instances")
    logger.info(
        "[watcher] inotify usage (%s): instances=%d watches=%d",
        context,
        instances,
        watches,
    )
    # Warn if we're within 10% of either limit — watchfiles fails silently
    # when notify runs out of watches, which manifests as "live reload
    # stopped working for some files".
    if max_watches and watches >= max_watches * 0.9:
        logger.warning(
            "[watcher] inotify watches=%d is ≥90%% of max_user_watches=%d — "
            "file changes in unwatched subtrees will be missed. "
            "Bump with: sudo sysctl fs.inotify.max_user_watches=524288",
            watches,
            max_watches,
        )
    if max_instances and instances >= max_instances * 0.9:
        logger.warning(
            "[watcher] inotify instances=%d is ≥90%% of max_user_instances=%d — "
            "bump with: sudo sysctl fs.inotify.max_user_instances=1024",
            instances,
            max_instances,
        )


# Event used to signal the multi-repo watcher to restart (e.g. when new
# repos are discovered).  The watcher thread checks this periodically.
_watcher_stop_event = threading.Event()


def signal_watcher_restart() -> None:
    """Signal the multi-repo watcher to restart with updated repo list."""
    _watcher_stop_event.set()


# Extensions we care about for live-reload
_WATCHED_EXTENSIONS = {".md"}

# Git internal files whose changes indicate repo state change (commit,
# merge, checkout, rebase, etc.).  Watching these lets us refresh the
# "recently changed" list after ``git commit`` even though no ``.md``
# file content actually changed on disk.
_GIT_STATE_FILES = {"index", "HEAD", "MERGE_HEAD", "REBASE_HEAD", "CHERRY_PICK_HEAD"}

# Quiet period: wait this long after last change before broadcasting,
# to coalesce rapid bursts like git branch switches.
_QUIET_PERIOD_S = 0.1
# Maximum time before forced broadcast even if changes keep arriving.
_MAX_WAIT_S = 1.0

# How often the watcher logs a heartbeat (counts of events seen, kept,
# dropped by extension, dropped by the vantage ignore filter).  This
# makes it possible to tell at a glance whether the inotify thread is
# alive but silent vs. truly dead.
_HEARTBEAT_INTERVAL_S = 60.0


@dataclass
class _WatcherStats:
    """Counters reported by the heartbeat logger.

    Reset at every heartbeat so the numbers describe the most recent
    interval, not the lifetime of the process.
    """

    events_total: int = 0
    kept: int = 0
    dropped_extension: int = 0  # not .md / not a tracked git state file
    dropped_ignore: int = 0  # matched .vantageignore / user ignore
    dropped_outside_repo: int = 0  # event under no watched root

    def reset(self) -> None:
        self.events_total = 0
        self.kept = 0
        self.dropped_extension = 0
        self.dropped_ignore = 0
        self.dropped_outside_repo = 0


class _GitAwareFilter(DefaultFilter):
    """Extends the default watchfiles filter to allow git state-file changes.

    ``DefaultFilter`` excludes the entire ``.git/`` tree.  We override
    ``__call__`` to let through a small set of top-level state files
    (e.g. ``index``, ``HEAD``) that change on commits, branch switches,
    rebases, etc.  The rest of ``.git/`` (objects, refs, logs, …) remains
    filtered to avoid noise.
    """

    @override
    def __call__(self, change: Change, path: str) -> bool:
        # Fast path: let the default filter handle non-.git paths
        parts = path.replace("\\", "/").split("/")
        # Check if any path component is ".git"
        try:
            git_idx = next(i for i, p in enumerate(parts) if p == ".git")
        except StopIteration:
            return super().__call__(change, path)
        # Allow .git/<state_file> (exactly one level deep); reject rest.
        return len(parts) == git_idx + 2 and parts[-1] in _GIT_STATE_FILES


def _classify(path: str, repo_root: Path | None = None) -> tuple[bool, str]:
    """Classify *path* for live-reload.

    Returns ``(keep, reason)`` where reason is one of:
      ``"kept"``               — relevant and not ignored
      ``"dropped_extension"``  — not .md, not a tracked git state file
      ``"dropped_ignore"``     — matched a vantage ignore rule
    """
    lower = path.lower()
    is_md = any(lower.endswith(ext) for ext in _WATCHED_EXTENSIONS)
    normalized = path.replace("\\", "/")
    parts = normalized.split("/")
    is_git_state = len(parts) >= 2 and parts[0] == ".git" and parts[-1] in _GIT_STATE_FILES

    if not (is_md or is_git_state):
        return False, "dropped_extension"

    if repo_root is not None:
        from vantage.services.ignore import get_matcher

        matcher = get_matcher(repo_root)
        if matcher.is_ignored(normalized):
            return False, "dropped_ignore"
    return True, "kept"


def _is_relevant(path: str, repo_root: Path | None = None) -> bool:
    """Back-compat wrapper around _classify for the async-side loop."""
    keep, _ = _classify(path, repo_root)
    return keep


def _is_git_state_change(path: str) -> bool:
    """Return True if the path is a git state file (not a .md content change)."""
    parts = path.replace("\\", "/").split("/")
    return len(parts) >= 2 and parts[0] == ".git" and parts[-1] in _GIT_STATE_FILES


async def _coalesce_and_broadcast(
    pending: set[str],
    repo_name: str | None = None,
) -> None:
    """Send a single batched message for accumulated paths."""
    if not pending:
        return

    # Always invalidate git-status cache on any file change — working
    # directory status reflects file state, not just git state.
    from vantage.services.fs_service import clear_md_dir_cache
    from vantage.services.git_service import clear_recent_files_cache, clear_status_cache

    clear_status_cache()

    # If a .md file was added or removed, clear the dir-has-markdown cache
    if any(p.lower().endswith(".md") for p in pending):
        clear_md_dir_cache()

    # If any pending path is a git state file, also invalidate the
    # recent-files cache so the next API call returns fresh data.
    has_git_change = any(_is_git_state_change(p) for p in pending)
    if has_git_change:
        clear_recent_files_cache()
        logger.debug("Cleared recent-files + git-status caches due to git state change")

    unique_paths = sorted(pending)
    msg: dict[str, object] = {"type": "files_changed", "paths": unique_paths}
    if repo_name:
        msg["repo"] = repo_name
        logger.info(f"Batch ({repo_name}): {len(unique_paths)} file(s) changed")
    else:
        logger.info(f"Batch: {len(unique_paths)} file(s) changed")
    logger.debug("Changed paths: %s", unique_paths)
    await manager.broadcast(msg)


async def watch_repo():
    """Watch single repo (legacy mode) with quiet-period coalescing.

    Uses the synchronous ``watch()`` in a daemon thread so that the
    (potentially slow) inotify initialization never blocks the event loop.
    """
    logger.info(f"Starting watcher for {settings.target_repo}")
    _log_inotify_limits()
    target = settings.target_repo.resolve()

    loop = asyncio.get_running_loop()
    queue: asyncio.Queue[set[tuple[Change, str]]] = asyncio.Queue()
    stats = _WatcherStats()

    def _run_sync_watcher() -> None:
        logger.info("Initializing file watcher...")
        t0 = time.monotonic()
        first = True
        try:
            for changes in watch(
                target,
                watch_filter=_GitAwareFilter(),
                ignore_permission_denied=True,
            ):
                if first:
                    logger.info(
                        "[startup] file watcher ready (%.0fms)", (time.monotonic() - t0) * 1000
                    )
                    _log_inotify_usage("after startup")
                    first = False
                # Drop ignored paths before queueing so they don't
                # flood the log or wake the async side for nothing.
                filtered: set[tuple[Change, str]] = set()
                for change, path in changes:
                    stats.events_total += 1
                    try:
                        rel = str(Path(path).relative_to(target))
                    except ValueError:
                        stats.dropped_outside_repo += 1
                        continue
                    keep, reason = _classify(rel, repo_root=target)
                    if not keep:
                        if reason == "dropped_extension":
                            stats.dropped_extension += 1
                        else:
                            stats.dropped_ignore += 1
                            # explain() is expensive (walks every ignore line);
                            # only build the reason string when DEBUG is on.
                            if logger.isEnabledFor(logging.DEBUG):
                                from vantage.services.ignore import get_matcher

                                explanation = get_matcher(target).explain(rel) or "?"
                                logger.debug(
                                    "[watcher] DROP %s %s (dropped_ignore:%s)",
                                    change.name,
                                    path,
                                    explanation,
                                )
                        continue
                    stats.kept += 1
                    logger.info("[watcher] %s %s", change.name, path)
                    filtered.add((change, path))
                if filtered:
                    loop.call_soon_threadsafe(queue.put_nowait, filtered)
        except Exception:
            logger.exception("[watcher] watch() thread crashed — live reload will stop")

    thread = threading.Thread(target=_run_sync_watcher, daemon=True)
    thread.start()

    async def _heartbeat() -> None:
        """Periodically log watcher stats so silent failures are visible."""
        while True:
            await asyncio.sleep(_HEARTBEAT_INTERVAL_S)
            alive = thread.is_alive()
            logger.info(
                "[watcher] heartbeat thread=%s events=%d kept=%d "
                "dropped_ext=%d dropped_ignore=%d dropped_outside=%d",
                "alive" if alive else "DEAD",
                stats.events_total,
                stats.kept,
                stats.dropped_extension,
                stats.dropped_ignore,
                stats.dropped_outside_repo,
            )
            if not alive:
                logger.warning("[watcher] inotify thread is not alive — live reload has stopped")
            _log_inotify_usage("heartbeat")
            stats.reset()

    asyncio.create_task(_heartbeat())

    pending: set[str] = set()
    quiet_task: asyncio.Task[None] | None = None
    batch_start: float | None = None

    async def flush() -> None:
        nonlocal batch_start
        paths = set(pending)
        pending.clear()
        batch_start = None
        await _coalesce_and_broadcast(paths)

    while True:
        changes = await queue.get()
        for _change, abs_path in changes:
            try:
                rel_path = str(Path(abs_path).relative_to(target))
            except ValueError:
                continue
            if _is_relevant(rel_path, repo_root=target):
                pending.add(rel_path)

        if not pending:
            continue

        now = asyncio.get_event_loop().time()
        if batch_start is None:
            batch_start = now

        if quiet_task and not quiet_task.done():
            quiet_task.cancel()

        if now - batch_start >= _MAX_WAIT_S:
            await flush()
        else:

            async def _delayed_flush() -> None:
                await asyncio.sleep(_QUIET_PERIOD_S)
                await flush()

            quiet_task = asyncio.create_task(_delayed_flush())


async def watch_multi_repo():
    """Watch multiple repos (daemon mode) with quiet-period coalescing.

    Uses the synchronous ``watch()`` in a daemon thread so that the
    (potentially slow) inotify initialization never blocks the event loop.

    When ``signal_watcher_restart()`` is called (e.g. after new repos are
    discovered), the current watch loop exits and restarts with the
    updated repo list from the daemon config.
    """
    daemon_config = get_daemon_config()
    if not daemon_config:
        await watch_repo()
        return

    _log_inotify_limits()

    loop = asyncio.get_running_loop()
    pending: dict[str, set[str]] = {}  # repo_name -> paths
    batch_start: float | None = None
    stats = _WatcherStats()
    current_thread_holder: dict[str, threading.Thread | None] = {"thread": None}

    async def flush() -> None:
        nonlocal batch_start
        snapshot = {k: set(v) for k, v in pending.items()}
        pending.clear()
        batch_start = None
        for repo_name, paths in snapshot.items():
            await _coalesce_and_broadcast(paths, repo_name)

    async def _heartbeat() -> None:
        while True:
            await asyncio.sleep(_HEARTBEAT_INTERVAL_S)
            t = current_thread_holder["thread"]
            alive = t.is_alive() if t else False
            logger.info(
                "[watcher] heartbeat thread=%s events=%d kept=%d "
                "dropped_ext=%d dropped_ignore=%d dropped_outside=%d",
                "alive" if alive else "DEAD",
                stats.events_total,
                stats.kept,
                stats.dropped_extension,
                stats.dropped_ignore,
                stats.dropped_outside_repo,
            )
            if t is not None and not alive:
                logger.warning("[watcher] inotify thread is not alive — live reload has stopped")
            _log_inotify_usage("heartbeat")
            stats.reset()

    asyncio.create_task(_heartbeat())

    while True:
        _watcher_stop_event.clear()

        watch_paths = []
        path_to_repo: dict[str, str] = {}
        for repo in daemon_config.repos:
            resolved = repo.path.resolve()
            watch_paths.append(resolved)
            path_to_repo[str(resolved)] = repo.name

        logger.info("Starting file watchers for %d repos", len(watch_paths))

        queue: asyncio.Queue[set[tuple[Change, str]] | None] = asyncio.Queue()

        def _start_watcher(
            paths: list[Path], q: asyncio.Queue[set[tuple[Change, str]] | None]
        ) -> None:
            logger.info("Initializing file watchers for %d repos...", len(paths))
            t0 = time.monotonic()
            first = True
            try:
                for changes in watch(
                    *paths,
                    watch_filter=_GitAwareFilter(),
                    stop_event=_watcher_stop_event,
                    ignore_permission_denied=True,
                ):
                    if first:
                        logger.info(
                            "[startup] file watchers ready (%.0fms)",
                            (time.monotonic() - t0) * 1000,
                        )
                        _log_inotify_usage("after startup")
                        first = False
                    # Pre-filter against each repo's ignore files so
                    # noisy subtrees (.yolo/, .worktrees/, …) never
                    # reach the async side or the journal.
                    filtered: set[tuple[Change, str]] = set()
                    for change, path in changes:
                        stats.events_total += 1
                        abs_obj = Path(path)
                        matched_root: Path | None = None
                        rel: str = ""
                        for root in paths:
                            try:
                                rel = str(abs_obj.relative_to(root))
                            except ValueError:
                                continue
                            matched_root = root
                            break
                        if matched_root is None:
                            stats.dropped_outside_repo += 1
                            logger.debug("[watcher] DROP %s (outside repos)", path)
                            continue
                        keep, reason = _classify(rel, repo_root=matched_root)
                        if not keep:
                            if reason == "dropped_extension":
                                stats.dropped_extension += 1
                            else:
                                stats.dropped_ignore += 1
                                if logger.isEnabledFor(logging.DEBUG):
                                    from vantage.services.ignore import get_matcher

                                    explanation = get_matcher(matched_root).explain(rel) or "?"
                                    logger.debug(
                                        "[watcher] DROP %s %s (dropped_ignore:%s)",
                                        change.name,
                                        path,
                                        explanation,
                                    )
                            continue
                        stats.kept += 1
                        logger.info("[watcher] %s %s", change.name, path)
                        filtered.add((change, path))
                    if filtered:
                        loop.call_soon_threadsafe(q.put_nowait, filtered)
            except Exception:
                logger.exception("[watcher] watch() thread crashed — live reload will stop")
            # Signal the async side that the watcher exited
            loop.call_soon_threadsafe(q.put_nowait, None)

        thread = threading.Thread(target=_start_watcher, args=(watch_paths, queue), daemon=True)
        thread.start()
        current_thread_holder["thread"] = thread

        quiet_task: asyncio.Task[None] | None = None
        restarting = False
        while True:
            changes = await queue.get()
            if changes is None:
                # Watcher was stopped — restart with updated repo list
                logger.info("File watcher stopped, restarting with updated repo list")
                restarting = True
                break

            for _change, abs_path in changes:
                abs_path_obj = Path(abs_path)
                for repo_path_str, name in path_to_repo.items():
                    repo_path = Path(repo_path_str)
                    try:
                        rel_path = str(abs_path_obj.relative_to(repo_path))
                    except ValueError:
                        continue
                    if _is_relevant(rel_path, repo_root=repo_path):
                        pending.setdefault(name, set()).add(rel_path)
                    break

            if not pending:
                continue

            now = asyncio.get_event_loop().time()
            if batch_start is None:
                batch_start = now

            if quiet_task and not quiet_task.done():
                quiet_task.cancel()

            if now - batch_start >= _MAX_WAIT_S:
                await flush()
            else:

                async def _delayed_flush() -> None:
                    await asyncio.sleep(_QUIET_PERIOD_S)
                    await flush()

                quiet_task = asyncio.create_task(_delayed_flush())

        if not restarting:
            break

File: internal/live/watcher.go
package live

import (
	"context"
	"errors"
	iofs "io/fs"
	"log/slog"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"

	fssvc "github.com/mschulkind-oss/vantage/internal/fs"
	gitsvc "github.com/mschulkind-oss/vantage/internal/git"
	"github.com/mschulkind-oss/vantage/internal/ignore"
	"github.com/mschulkind-oss/vantage/internal/review"
)

const (
	// quietPeriod is how long the watcher waits after the last change before
	// broadcasting, to coalesce rapid bursts (e.g. a branch switch).
	quietPeriod = 100 * time.Millisecond
	// maxWait caps the total batch window so a steady stream of changes still
	// flushes within a bounded time.
	maxWait = 1 * time.Second
	// heartbeatInterval is how often watcher stats are logged so a silently
	// dead inotify thread is visible in the journal.
	heartbeatInterval = 60 * time.Second
)

// gitStateFiles are the one-level .git files whose changes signal a repo-state
// change (commit, merge, checkout, rebase) even when no .md content changed.
var gitStateFiles = map[string]struct{}{
	"index":            {},
	"HEAD":             {},
	"MERGE_HEAD":       {},
	"REBASE_HEAD":      {},
	"CHERRY_PICK_HEAD": {},
}

// classify decides whether a repo-relative path (slash-separated) is relevant
// for live reload. It returns keep=true for Markdown files and for one-level
// .git state files, and isGitState=true only for the latter. Ignore filtering
// is applied separately by the Watcher (classify is pure for testability).
func classify(rel string) (keep bool, isGitState bool) {
	norm := filepath.ToSlash(rel)
	parts := strings.Split(norm, "/")

	// Only one-level .git state files (.git/<state>) are relevant; deeper
	// paths like .git/logs/HEAD are pruned from the watch set and never kept.
	if len(parts) == 2 && parts[0] == ".git" {
		_, ok := gitStateFiles[parts[1]]
		return ok, ok
	}
	if strings.HasSuffix(strings.ToLower(norm), ".md") {
		return true, false
	}
	return false, false
}

// shouldPruneDir reports whether a directory (repo-relative, slash-separated)
// should be excluded from the recursive watch set. The .git subtree is pruned
// except for its immediate top level (so one-level state files remain watched);
// .vantage is pruned except for its inbox (so review deliveries generate
// events even though the dir is always-ignored elsewhere); ignored/excluded
// directories are pruned via the ignore matcher.
func shouldPruneDir(rel string, matcher *ignore.Matcher) bool {
	norm := filepath.ToSlash(rel)
	if norm == "." || norm == "" {
		return false
	}
	parts := strings.Split(norm, "/")
	// Keep ".git" itself (depth 1) so its state files generate events; prune
	// anything deeper under .git (objects, refs, logs, …).
	if parts[0] == ".git" {
		return len(parts) > 1
	}
	// Keep ".vantage" and its inbox — checked before the matcher, which
	// always-ignores the dir. Anything deeper is vantage-owned state the
	// watcher has no business in.
	if parts[0] == ".vantage" {
		return len(parts) > 1 && norm != review.InboxRel
	}
	if matcher != nil && matcher.IsIgnored(norm, true) {
		return true
	}
	return false
}

// isInboxPath reports whether a repo-relative slash path is the review inbox
// directory or anything inside it. Inbox events trigger consumption instead of
// the files_changed pipeline (classify would drop them: they are not Markdown).
func isInboxPath(rel string) bool {
	return rel == review.InboxRel || strings.HasPrefix(rel, review.InboxRel+"/")
}

// Watcher recursively watches a repository for Markdown and git-state changes,
// coalesces them, invalidates caches, consumes review inbox deliveries, and
// broadcasts files_changed messages through a Manager.
type Watcher struct {
	root     string
	repoName string

	manager *Manager
	store   *review.Store
	matcher *ignore.Matcher
	logger  *slog.Logger

	fsw *fsnotify.Watcher

	mu    sync.Mutex
	stats watcherStats
}

// watcherStats are reset every heartbeat so they describe the most recent
// interval rather than the lifetime of the process.
type watcherStats struct {
	eventsTotal    int
	kept           int
	droppedExt     int
	droppedIgnore  int
	droppedOutside int
}

// NewWatcher constructs a Watcher rooted at the repository at root. repoName is
// echoed in broadcast payloads (empty for single-repo mode). store consumes
// review inbox deliveries and backs the stale-changelog warning; it may be nil
// to disable both. useIgnoreFiles toggles .vantageignore/user-ignore pruning.
// If logger is nil the default slog logger is used.
func NewWatcher(root, repoName string, mgr *Manager, store *review.Store, useIgnoreFiles bool, logger *slog.Logger) (*Watcher, error) {
	if logger == nil {
		logger = slog.Default()
	}
	abs, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	return &Watcher{
		root:     filepath.Clean(abs),
		repoName: repoName,
		manager:  mgr,
		store:    store,
		matcher:  ignore.GetMatcher(abs, useIgnoreFiles),
		logger:   logger.With("component", "watcher", "repo", repoName),
	}, nil
}

// Start begins watching. It adds the recursive watch set, then runs the event
// loop until ctx is cancelled or Close is called. Start blocks; run it in its
// own goroutine.
func (w *Watcher) Start(ctx context.Context) error {
	fsw, err := fsnotify.NewWatcher()
	if err != nil {
		return err
	}
	w.fsw = fsw

	added := w.addRecursive(w.root)
	w.logger.Info("watcher started", "root", w.root, "watched_dirs", added)

	// Deliveries that landed while the server was down are consumed before the
	// first event can arrive.
	w.consumeInbox()

	out := make(chan []string, 1)
	co := newCoalescer(quietPeriod, maxWait, func(paths []string) { out <- paths })

	heartbeat := time.NewTicker(heartbeatInterval)
	defer heartbeat.Stop()

	for {
		select {
		case <-ctx.Done():
			co.stop()
			_ = w.fsw.Close()
			return ctx.Err()

		case paths := <-out:
			w.flush(paths)

		case ev, ok := <-w.fsw.Events:
			if !ok {
				co.stop()
				return nil
			}
			w.handleEvent(ev, co)

		case err, ok := <-w.fsw.Errors:
			if !ok {
				continue
			}
			w.handleError(err)

		case <-heartbeat.C:
			w.logHeartbeat()
		}
	}
}

// Close stops watching and releases the inotify handle. It is safe to call once
// Start has returned or to unblock a running Start (which also closes on ctx
// cancellation).
func (w *Watcher) Close() error {
	if w.fsw == nil {
		return nil
	}
	return w.fsw.Close()
}

// addRecursive walks dir and adds a watch for every surviving directory,
// pruning the .git subtree (except its top level) and ignored directories.
// fsnotify is not recursive, so each directory is added individually. Returns
// the number of directories added.
func (w *Watcher) addRecursive(dir string) int {
	added := 0
	_ = filepath.WalkDir(dir, func(path string, d iofs.DirEntry, err error) error {
		if err != nil {
			// Unreadable entries (permissions, races) are skipped, not fatal.
			if d != nil && d.IsDir() {
				return iofs.SkipDir
			}
			return nil
		}
		if !d.IsDir() {
			return nil
		}
		rel, relErr := filepath.Rel(w.root, path)
		if relErr != nil {
			return nil
		}
		if shouldPruneDir(rel, w.matcher) {
			return iofs.SkipDir
		}
		if addErr := w.fsw.Add(path); addErr != nil {
			w.logger.Debug("watcher: failed to add watch", "path", path, "error", addErr)
			return nil
		}
		added++
		return nil
	})
	return added
}

// handleEvent classifies an fsnotify event and either feeds the coalescer a
// kept repo-relative path or, for newly created directories, extends the watch
// set. Removed paths are dropped from accounting by classify.
func (w *Watcher) handleEvent(ev fsnotify.Event, co *coalescer) {
	w.mu.Lock()
	w.stats.eventsTotal++
	w.mu.Unlock()

	// Newly created directories must be added to the (non-recursive) watch set.
	if ev.Has(fsnotify.Create) {
		if info, err := os.Stat(ev.Name); err == nil && info.IsDir() {
			rel, relErr := filepath.Rel(w.root, ev.Name)
			if relErr == nil && !shouldPruneDir(rel, w.matcher) {
				w.addRecursive(ev.Name)
				// A .vantage or inbox dir appearing after startup may already
				// hold deliveries written before its watch existed.
				if norm := filepath.ToSlash(rel); norm == ".vantage" || isInboxPath(norm) {
					w.consumeInbox()
				}
			}
			return
		}
	}

	rel, err := filepath.Rel(w.root, ev.Name)
	if err != nil || strings.HasPrefix(rel, "..") {
		w.mu.Lock()
		w.stats.droppedOutside++
		w.mu.Unlock()
		return
	}
	rel = filepath.ToSlash(rel)

	// Inbox traffic is consumed immediately; it never enters the coalesced
	// files_changed flow. Consumption's own renames and deletes re-trigger
	// this path, each a cheap pass over an empty directory.
	if isInboxPath(rel) {
		w.mu.Lock()
		w.stats.kept++
		w.mu.Unlock()
		w.logger.Debug("watcher inbox event", "op", ev.Op.String(), "path", rel)
		w.consumeInbox()
		return
	}

	keep, _ := classify(rel)
	if !keep {
		w.mu.Lock()
		w.stats.droppedExt++
		w.mu.Unlock()
		return
	}
	// Ignore filtering applies to .md content paths; .git state files are
	// never user-ignored.
	if w.matcher != nil && !strings.HasPrefix(rel, ".git/") && w.matcher.IsIgnored(rel, false) {
		w.mu.Lock()
		w.stats.droppedIgnore++
		w.mu.Unlock()
		return
	}

	w.mu.Lock()
	w.stats.kept++
	w.mu.Unlock()
	w.logger.Debug("watcher event kept", "op", ev.Op.String(), "path", rel)
	co.add(rel)
}

// flush processes a coalesced change set: invalidate caches, warn about saved
// documents still carrying retired-protocol changelog blocks, and broadcast
// files_changed.
func (w *Watcher) flush(paths []string) {
	if len(paths) == 0 {
		return
	}

	gitsvc.ClearStatusCache()

	hasMarkdown := false
	hasGitState := false
	for _, p := range paths {
		if strings.HasSuffix(strings.ToLower(p), ".md") {
			hasMarkdown = true
		}
		if _, isState := classify(p); isState {
			hasGitState = true
		}
	}
	if hasMarkdown {
		fssvc.ClearMarkdownDirCache()
	}
	if hasGitState {
		gitsvc.ClearRecentFilesCache()
		w.logger.Debug("cleared recent-files cache due to git state change")
	}

	if w.store != nil {
		for _, rel := range paths {
			if !strings.HasSuffix(strings.ToLower(rel), ".md") {
				continue
			}
			content, err := os.ReadFile(filepath.Join(w.root, filepath.FromSlash(rel)))
			if err != nil {
				continue
			}
			// The changelog protocol is retired: a marker in a saved document
			// means an agent is following a stale clipboard payload, and its
			// response was NOT recorded. Warn loudly (log + UI notice) instead of
			// silently reproducing the old lost-turn failure. Once per save.
			if review.ContainsChangelogBlock(string(content)) {
				w.logger.Warn("review: document contains a retired-protocol changelog block; "+
					"the agent's response was NOT recorded — copy the comments again to hand it "+
					"the current instructions", "path", rel)
				w.manager.Broadcast(changelogIgnoredMessage{Type: "changelog_ignored", Repo: w.repoName, Path: rel})
			}
		}
	}

	sorted := append([]string(nil), paths...)
	sort.Strings(sorted)
	msg := filesChangedMessage{Type: "files_changed", Paths: sorted}
	if w.repoName != "" {
		msg.Repo = w.repoName
	}
	w.logger.Info("files changed", "count", len(sorted))
	w.manager.Broadcast(msg)
}

// filesChangedMessage is the broadcast payload for a coalesced change set.
type filesChangedMessage struct {
	Type  string   `json:"type"`
	Repo  string   `json:"repo,omitempty"`
	Paths []string `json:"paths"`
}

// reviewChangedMessage mirrors the server's review_changed push so inbox
// deliveries and API commands look identical to the browser. Repo is
// intentionally not omitempty: the frontend matches it against its current
// repo, and the single-repo sentinel is the empty string, not an absent key.
type reviewChangedMessage struct {
	Type string `json:"type"`
	Repo string `json:"repo"`
	Path string `json:"path"`
}

// changelogIgnoredMessage is the stale-payload warning push: a saved document
// still carries a retired-protocol changelog block, so whatever the agent
// wrote there was ignored. Same repo semantics as reviewChangedMessage.
type changelogIgnoredMessage struct {
	Type string `json:"type"`
	Repo string `json:"repo"`
	Path string `json:"path"`
}

// consumeInbox drains the repo's review inbox and pushes review_changed for
// every document a delivery mutated, so open browsers reload that review. It
// runs at startup and on every event under the inbox; a missing or empty
// inbox is a cheap no-op.
func (w *Watcher) consumeInbox() {
	if w.store == nil {
		return
	}
	for _, p := range w.store.ConsumeInbox(w.root, w.repoName) {
		w.logger.Info("review: consumed inbox delivery", "path", p)
		w.manager.Broadcast(reviewChangedMessage{Type: "review_changed", Repo: w.repoName, Path: p})
	}
}

// handleError surfaces an fsnotify error, adding a hint for the common inotify
// watch-exhaustion case (ENOSPC) which otherwise manifests as live reload
// silently failing for some files.
func (w *Watcher) handleError(err error) {
	if errors.Is(err, fsnotify.ErrEventOverflow) || strings.Contains(err.Error(), "no space left") {
		w.logger.Error("watcher: inotify watch limit reached — live reload will miss some changes; "+
			"raise fs.inotify.max_user_watches (e.g. sysctl fs.inotify.max_user_watches=524288)",
			"error", err)
		return
	}
	w.logger.Warn("watcher error", "error", err)
}

// logHeartbeat emits and resets the interval stats so a stalled watcher is
// distinguishable from a quiet one.
func (w *Watcher) logHeartbeat() {
	w.mu.Lock()
	s := w.stats
	w.stats = watcherStats{}
	w.mu.Unlock()
	w.logger.Info("watcher heartbeat",
		"events", s.eventsTotal,
		"kept", s.kept,
		"dropped_ext", s.droppedExt,
		"dropped_ignore", s.droppedIgnore,
		"dropped_outside", s.droppedOutside,
	)
}
